# `setTimeout` can not `return` value ITSELF!!
- https://stackoverflow.com/questions/24928846/get-return-value-from-settimeout?fbclid=IwAR1_fKA_nOC27Q0qvcS8CbvRiQMi63GuOTqqiDa4TTAoSZ7kO-MKrnWv5x8
- when I tested async/await code with setTimeout, I get this
```js
function x () {
    const y = setTimeout(function () {
        return 'done';
    }, 1000);
    return y;
}

console.log(x());

// 2
```

## Reason
- The function that called `setTimeout` will **finish** executing and return before the function you pass to `setTimeout` is even called.
- 因為呼叫setTimeout時，setTimeout本身馬上就執行了，此時callback根本還沒執行，所以不可能return! 又一個呼應event loop的例子。

# have to return by using `Promise`, `resolve()`
- 如果想要用setTimeout回傳東西，需要包裝成一個Promise，指定到一個變數裡：Promise傳入 `(response, reject) => { }的function` ，在這裡面去用response包住setTimeout去return東西。
- 直接宣告一個promise到變數proObj裡面，並且用promise帶入的resolve()將setTimeout裡面的東西回傳
- 或是宣告一function return上述的promise

```js
// 方法一：直接宣告一個promise到變數proObj裡面，
//        並且用promise帶入的resolve()將setTimeout裡面的東西回傳

const proObj = new Promise((resolve, reject) => {
     setTimeout(() => 
       resolve('done!');
     );
   });

x().then((done) => {
  console.log(done); // --> 'done!'
});

// 方法二：用function包住，並return 方法一的promise
```

## promise語法複習：

```js
const bigPro = new Promise((resolve, reject) => resolve("alert!!!"))
```

